File: src/main/java/org/dxworks/utils/java/rest/client/RestClient.java
package org.dxworks.utils.java.rest.client;

import com.google.api.client.http.HttpRequestInitializer;
import lombok.experimental.Delegate;

public abstract class RestClient {

    protected HttpClient httpClient;

    @Delegate
    protected PathResolver pathResolver;

    public RestClient(String apiBaseUrl) {
        this.httpClient = new HttpClient();
        this.pathResolver = new PathResolver(apiBaseUrl);
    }

    public RestClient(String apiBaseUrl, HttpRequestInitializer httpRequestInitializer) {
        this.httpClient = new HttpClient(httpRequestInitializer);
        this.pathResolver = new PathResolver(apiBaseUrl);
    }

    public RestClient(String apiBaseUrl, HttpClient httpClient) {
        this.httpClient = httpClient;
        this.pathResolver = new PathResolver(apiBaseUrl);
    }
}

File: src/main/kotlin/org/dxworks/utils/java/rest/client/RestClient.kt
package org.dxworks.utils.java.rest.client

import com.google.api.client.http.HttpRequestInitializer

abstract class RestClient private constructor(
    private val pathResolver: PathResolver,
    protected val httpClient: HttpClient
) {
    constructor(apiBaseUrl: String)
            : this(PathResolver(apiBaseUrl), HttpClient())

    constructor(apiBaseUrl: String, httpRequestInitializer: HttpRequestInitializer?)
            : this(PathResolver(apiBaseUrl), HttpClient(httpRequestInitializer))

    constructor(apiBaseUrl: String, httpClient: HttpClient)
            : this(PathResolver(apiBaseUrl), httpClient)

    open fun getApiPath(variableValues: Map<String, String>, vararg pathVariables: String): String {
        return pathResolver.getApiPath(variableValues, *pathVariables)
    }

    open fun replacePlaceholders(v: String, variablesValues: Map<String, String>): String {
        return pathResolver.replacePlaceholders(v, variablesValues)
    }

    open fun getApiPath(vararg pathVariables: String): String {
        return pathResolver.getApiPath(*pathVariables)
    }
}
